Convert mdns Builder classes to use PacketBufferHandle

kpschoedel opened the pull request:
#### Problem

Code should use `PacketBufferHandle` rather than `PacketBuffer`
directly.

#### Summary of Changes

Convert mdns::Minimal::QueryBuilder and mdns::Minimal::ResponseBuilder
to take and hold PacketBufferHandle.

Part of issue #2707 - Figure out a way to express PacketBuffer ownership in the type system

bzbarsky-apple commented on examples/minimal-mdns/client.cpp:
```diff
@@ -213,7 +213,7 @@ void BroadcastPacket(mdns::Minimal::ServerBase * server)
     query.Split(gOptions.query);
 
-    mdns::Minimal::QueryBuilder builder(buffer.Get_ForNow());
+    mdns::Minimal::QueryBuilder builder(buffer.Retain());
```
Hmm.  It seems like we could make the builder take a const ref to PacketBufferHandle (and have its member be a const ref) and avoid the extra refcounting  And here we could scope it so it goes out of scope before we do the `BroadcastSend` call...

kpschoedel replied:
Ah, yes. I probably had TLV's chaining in mind when I converted this one. Fixed.

bzbarsky-apple commented:
@kpschoedel Merge conflicts?

andy31415 commented:
also #4102 has some conversions to PacketBufferHandle (not sure if the same, but if we could make them similar it may be easier to merge)

kpschoedel commented:
> also #4102 has some conversions to PacketBufferHandle (not sure if the same, but if we could make them similar it may be easier to merge)

Yes, I think #4102 makes the ResponseBuilder part of this PR redundant, so I'll just remove it.

kpschoedel commented:
Converted to draft as this now depends on #4102 

Removed ResponseBuilder changes (redundant after #4102) and converted QueryBuilder to use the same interface pattern as ResponseBuilder there.
